File: LeetCode/Easy/Minimum Depth of Tree/Go/minimum_depth.go
package main

import "fmt"

type TreeNode struct {
	Val         int
	Left, Right *TreeNode
}

func New(val int) *TreeNode { return &TreeNode{Val: val, Left: nil, Right: nil} }

type Pair struct {
	node  *TreeNode
	depth int
}

func minimum_depth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	queue := []Pair{{node: root, depth: 1}}
	depth := 0
	for len(queue) != 0 {
		curr := queue[0].node
		depth = queue[0].depth
		queue = queue[1:]
		if curr.Right == nil && curr.Left == nil {
			return depth
		}
		if curr.Left != nil {
			queue = append(queue, Pair{node: curr.Left, depth: depth + 1})
		}
		if curr.Right != nil {
			queue = append(queue, Pair{node: curr.Right, depth: depth + 1})
		}
	}
	return depth
}

func main() {
	root := New(3)
	root.Left = New(9)
	root.Right = New(20)
	root.Right.Left = New(15)
	root.Right.Right = New(7)
	fmt.Println(minimum_depth(root))
}

File: LeetCode/Easy/Minimum Depth of Tree/Java/minimumDepth.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int val;
    public int minDepth(TreeNode root) {
        if (root == null) return 0;
        return recur(root);
    }

    private int recur(TreeNode root) {
        if (root == null) return Integer.MAX_VALUE;
        if (root.left == null && root.right == null) return 1;
        int left = recur(root.left);
        int right = recur(root.right);
        return 1+Math.min(left, right);
    }
}
